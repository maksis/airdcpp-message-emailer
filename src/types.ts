export type Message = {
  from: { nick: string };
  text: string;
  time: number; // seconds epoch
};

export type SessionId = string | number;

export type SessionInfoGetter = (id: SessionId) => Promise<SessionInfo>;

export type SessionBase = {
  // id: SessionId;
  message_counts: {
    unread: {
      bot: number;
      status: number;
      user: number;
    };
  };
};

export type HubSession = SessionBase & {
  identity: {
    name: string;
  };
  user?: {
    nicks?: string;
  };
};

export type PrivateChatSession = SessionBase & {
  user: {
    nicks: string;
  };
  identity?: {
    name?: string;
  };
};

export type SessionInfo = HubSession | PrivateChatSession;

export type SessionNameGetter = (sessionInfo: SessionInfo) => string;

export type SessionType = {
  title: string;
  sessionNameGetter: SessionNameGetter;
};

export type MessageCache = Record<string, Message[]>;
