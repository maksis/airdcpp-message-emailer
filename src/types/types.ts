import { ChatMessage, SessionBase, SessionId, SessionInfo } from "./api";

export type SessionInfoFetcher<SessionType extends SessionBase> = (id: SessionId, skipCache?: boolean) => Promise<SessionType>;
export type SessionNameGetter = (sessionInfo: SessionInfo) => string;

export type SessionParser = {
  title: string;
  sessionNameGetter: SessionNameGetter;
};

export type MessageCache = Record<string, ChatMessage[]>;
