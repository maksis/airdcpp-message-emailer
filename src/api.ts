import { APISocket } from 'airdcpp-apisocket';
import { SessionInfoFetcher } from './types/types';
import { HubSession, PrivateChatSession, SeverityEnum } from './types';


export const API = (socket: APISocket) => {
  const getPrivateChatSession: SessionInfoFetcher<PrivateChatSession> = (sessionId) => socket.get(`private_chat/${sessionId}`);
  const getHubSession: SessionInfoFetcher<HubSession> = (sessionId) => socket.get(`hubs/${sessionId}`);


  const postEvent = async (text: string, severity: SeverityEnum) => {
    return socket.post(
      'events',
      {
        text,
        severity,
      }
    );
  };

  return {
    getPrivateChatSession,
    getHubSession,
    postEvent,
  }
};

export type APIType = ReturnType<typeof API>;
