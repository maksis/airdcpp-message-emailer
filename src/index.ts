import 'source-map-support/register';
import { ManagedExtension } from 'airdcpp-extension';
import init from './main';

// Entry point that is executed by the extension manager
//
// The file isn't executed when running development server so it shouldn't 
// generally contain any extension-specific code

// See https://github.com/airdcpp-web/airdcpp-extension-js for usage information
ManagedExtension(init, {
  // Possible custom options for airdcpp-apisocket can be listed here
});
