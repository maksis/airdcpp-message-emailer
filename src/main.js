'use strict';

import nodemailer from 'nodemailer';

import MessageBuilder from './MessageBuilder';
import SessionTypes from './SessionTypes';

import { DEFAULT_TO, SettingDefinitions } from './SettingDefinitions';
const CONFIG_VERSION = 1;

const SettingsManager = require('airdcpp-extension-settings');

export default function (socket, extension) {
	const settings = SettingsManager(socket, {
		extensionName: extension.name, 
		configFile: extension.configPath + 'config.json',
		configVersion: CONFIG_VERSION,
		definitions: SettingDefinitions,
	});

	let transporter;
	let flushTimeout;
	let hostname;

	const cache = {
		privateMessages: {},
		hubMessages: {},
	};

	const hasConfig = () => settings.getValue('to') !== DEFAULT_TO;

	const constructSmtpSettings = () => {
		const ret = {
			host: settings.getValue('smtp_host'),
			port: settings.getValue('smtp_port'),
			secure: settings.getValue('smtp_secure'),
		}

		const user = settings.getValue('smtp_user');
		const pass = settings.getValue('smtp_password');
		if (user && pass) {
			ret['auth'] = {
				user,
				pass,
			}
		}

		return ret;
	};

	const constructMail = (text) => {
		return {
				from: settings.getValue('from'),
				to: settings.getValue('to'),
				subject: `Chat summary from AirDC++ (${hostname})`,
				text,
		};
	};

	const privateChatInfoGetter = sessionId => socket.get(`private_chat/${sessionId}`);
	const hubInfoGetter = sessionId => socket.get(`hubs/${sessionId}`);

	// Send cached messages (if there are any)
	const flushCache = async () => {
		socket.logger.verbose('Flushing cache...');
		let messageSummary = '';
		messageSummary += await MessageBuilder.constructSummary(SessionTypes.privateChat, cache.privateMessages, privateChatInfoGetter);
		messageSummary += await MessageBuilder.constructSummary(SessionTypes.hub, cache.hubMessages, hubInfoGetter);

		if (messageSummary) {
			transporter.sendMail(constructMail(messageSummary), (error, info) => {
				if (error) {
					socket.logger.error(`Failed to send email: ${error}`);
					socket.post('events', {
						text: `${extension.name}: failed to email message summary (see the extension error log for more information)`,
						severity: 'error',
					});
				} else {
					socket.logger.info(`Message ${info.messageId} sent: ${info.response}`);
				}
			});
		} else {
			socket.logger.verbose('Nothing to send');
		}

		cache.privateMessages = {};
		cache.hubMessages = {};
		flushTimeout = undefined;
	};

	const onIncomingMessage = (cacheKey, message, sessionId) => {
		if (!hasConfig() || !settings.getValue('flush_interval')) {
			socket.logger.verbose('Caching disabled due to current configuration');
			return;
		}
		
		const sessionCache = cache[cacheKey];
		sessionCache[sessionId] = sessionCache[sessionId] || [];
		sessionCache[sessionId].push(message);

		if (!flushTimeout) {
			const flushMs = settings.getValue('flush_interval') * 60 * 1000;
			socket.logger.verbose('Scheduling flush', flushMs);
			flushTimeout = setTimeout(flushCache, flushMs);
		}
	};

	const isSmtpSetting = key => key.indexOf('smtp_') === 0;

	extension.onStart = async (sessionInfo) => {
		hostname = sessionInfo.system_info.hostname;
		settings.onValuesUpdated = updatedValues => {
			if (Object.keys(updatedValues).find(isSmtpSetting)) {
				socket.logger.verbose('Creating mailer interface');
				try {
					transporter = nodemailer.createTransport(constructSmtpSettings());
				} catch (e) {
					socket.logger.error(`Failed to initialize mailer interface: ${e}`);
					process.exit(1);
				}

				if (hasConfig()) {
					transporter.verify((error, success) => {
						socket.logger.verbose('Verifying settings');
						if (error) {
							socket.post('events', {
								text: `${extension.name}: SMTP setting validation failed (${error})`,
								severity: 'error',
							});
						} else {
							socket.logger.verbose('Setting validation succeeded');
							/*socket.post('events', {
								text: `${extension.name}: SMTP settings were validated successfully`,
								severity: 'info',
							});*/
						}
					});
				}
			}
		};

		await settings.load();
		socket.logger.verbose('Settings were loaded');

		if (settings.getValue('send_hub_logs')) {
			socket.addListener('hubs', 'hub_message', onIncomingMessage.bind(this, 'hubMessages'));
		}

		if (settings.getValue('send_private_logs')) {
			socket.addListener('private_chat', 'private_chat_message', onIncomingMessage.bind(this, 'privateMessages'));
		}
	};

	extension.onStop = () => {
		clearTimeout(flushTimeout);
	};
};