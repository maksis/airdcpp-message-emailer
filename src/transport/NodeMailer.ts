import nodemailer from 'nodemailer';

import { Context, SeverityEnum } from '../types';
import { hasConfig } from '../utils';

export const NodeMailer = (context: Context) => {
  const { api, logger, getExtSetting, extension } = context;

  let hostname: string;
  let transporter: nodemailer.Transporter | undefined;

  const constructSmtpSettings = () => {
    const ret = {
      host: getExtSetting('smtp_host'),
      port: getExtSetting('smtp_port'),
      secure: getExtSetting('smtp_secure'),
    };

    const user = getExtSetting('smtp_user');
    const pass = getExtSetting('smtp_password');
    if (user && pass) {
      ret['auth'] = {
        user,
        pass,
      };
    }

    return ret;
  };

  const constructMail = (text: string) => {
    const mailOptions: nodemailer.SendMailOptions = {
      from: getExtSetting('from'),
      to: getExtSetting('to'),
      subject: `Chat summary from AirDC++ (${hostname})`,
      text,
    };

    return mailOptions;
  };

  const createTransporter = () => {
    const smtpSettings = constructSmtpSettings();
    return nodemailer.createTransport(smtpSettings);
  };

  const sendMail = (messageSummary: string) => {
    transporter!.sendMail(constructMail(messageSummary), (error: any, info: any) => {
      if (error) {
        logger.error(`Failed to send email: ${error}`);
        api.postEvent(`${extension.name}: failed to email message summary (see the extension error log for more information)`, SeverityEnum.ERROR);
      } else {
        logger.info(`Message ${info.messageId} sent: ${info.response}`);
      }
    });
  }

  
  const reset = (newHostName: string) => {
    hostname = newHostName;

    logger.verbose('Creating mailer interface');
    try {
      transporter = nodemailer.createTransport(constructSmtpSettings());
    } catch (e) {
      logger.error(`Failed to initialize mailer interface: ${e}`);
      process.exit(1);
    }

    if (hasConfig(getExtSetting)) {
      transporter.verify((error: any, _success: boolean) => {
        logger.verbose('Verifying settings');
        if (error) {
          api.postEvent(`${extension.name}: SMTP setting validation failed (${error})`, SeverityEnum.ERROR);
        } else {
          logger.verbose('Setting validation succeeded');
        }
      });
    }
  }

  return {
    createTransporter,
    sendMail,
    reset,
  };
};

export type Transporter = ReturnType<typeof NodeMailer>;
