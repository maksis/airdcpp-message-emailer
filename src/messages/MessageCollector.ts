import MessageBuilder from './MessageBuilder';
import { Context, ChatMessage, MessageCache, SessionId, PrivateChatSession, HubSession } from '../types';

import { SessionPropertyParsers } from './SessionPropertyParsers';

import { Transporter } from '../transport/NodeMailer';
import { hasConfig } from '../utils';
import { parseHubMessageLevel, parsePrivateMessageLevel } from './MessageLevelParser';
import { HubMessageModePriority, PrivateChatModePriority } from '../settings';
import { SessionCache } from './SessionCache';

export const MessageCollector = (mailer: Transporter, context: Context) => {
  const { api, logger, getExtSetting } = context;

  let flushTimeout: NodeJS.Timeout | undefined;

  const sessionCache = {
    privateMessages: SessionCache<PrivateChatSession>(api.getPrivateChatSession),
    hubMessages: SessionCache<HubSession>(api.getHubSession),
  };

  const sessionMessageCache = {
    privateMessages: {} as MessageCache,
    hubMessages: {} as MessageCache,
  };

  // Send cached messages (if there are any)
  const flushCache = async () => {
    logger.verbose('Flushing cache...');
    let messageSummary = '';
    messageSummary += await MessageBuilder.constructSummary(SessionPropertyParsers.privateChat, sessionMessageCache.privateMessages, sessionCache.privateMessages.getSession);
    messageSummary += await MessageBuilder.constructSummary(SessionPropertyParsers.hub, sessionMessageCache.hubMessages, sessionCache.hubMessages.getSession);

    if (messageSummary) {
      mailer.sendMail(messageSummary);
    } else {
      logger.verbose('Nothing to send');
    }

    sessionMessageCache.privateMessages = {};
    sessionMessageCache.hubMessages = {};
    flushTimeout = undefined;
  };

  const onIncomingMessage = (cacheKey: keyof typeof sessionMessageCache, message: ChatMessage, sessionId: SessionId) => {
    if (!hasConfig(getExtSetting) || !getExtSetting('flush_interval')) {
      logger.verbose('Caching disabled due to current configuration');
      return;
    }

    const sessionMessages = sessionMessageCache[cacheKey];
    sessionMessages[sessionId] = sessionMessages[sessionId] || [];
    sessionMessages[sessionId].push(message);
    logger.verbose(`Cached message ${message.id} for session ${sessionId}, total: ${sessionMessages[sessionId].length}`);

    if (!flushTimeout) {
      const flushMs = getExtSetting('flush_interval') * 60 * 1000;
      logger.verbose('Scheduling flush', flushMs);
      flushTimeout = setTimeout(flushCache, flushMs);
    }
  };

  const onHubMessage = async (message: ChatMessage, sessionId: SessionId) => {
    try {
      const hubSession = await sessionCache.hubMessages.getSession(sessionId);

      const level = HubMessageModePriority[parseHubMessageLevel(message, hubSession.settings.use_main_chat_notify)];
      const requiredLevel = HubMessageModePriority[getExtSetting('hub_message_mode')];
      if (level < requiredLevel) {
        logger.verbose(`Ignoring hub message due to current configuration (has level ${level} while ${requiredLevel} is required)`);
        return;
      }
    } catch (error) {
      logger.error('Could not fetch hub session for a received message', error);
      return;
    }

    onIncomingMessage('hubMessages', message, sessionId);
  }

  const onPrivateMessage = async (message: ChatMessage, sessionId: SessionId) => {
    const level = PrivateChatModePriority[parsePrivateMessageLevel(message)];
    const requiredLevel = PrivateChatModePriority[getExtSetting('private_chat_message_mode')];
    if (level < requiredLevel) {
      logger.verbose(`Ignoring private message due to current configuration (has level ${level} while ${requiredLevel} is required)`);
      return;
    }

    onIncomingMessage('privateMessages', message, sessionId);
  };

  return {
    onHubMessage,
    onPrivateMessage,
    stop: () => {
      clearTimeout(flushTimeout);
    },

    getPrivateChatSessionMessages: (sessionId: SessionId) => {
      return sessionMessageCache.privateMessages[sessionId];
    },
    getHubSessionMessages: (sessionId: SessionId) => {
      return sessionMessageCache.hubMessages[sessionId];
    },
    hasPendingFlush: () => {
      return !!flushTimeout;
    },
    flush: () => {
      return flushCache();
    }
  };
};