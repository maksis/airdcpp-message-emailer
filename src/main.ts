import { APISocket } from 'airdcpp-apisocket';

import SettingsManager from 'airdcpp-extension-settings';
import { ExtensionEntryData } from 'airdcpp-extension';

import { HubMessageCollectorMode, PrivateChatCollectorMode, SettingDefinitions } from './settings';

import { MessageCollector } from './messages';
import { NodeMailer } from './transport';

import { Context } from './types';
import { API } from './api';

const CONFIG_VERSION = 1;

export default function init(socket: APISocket, extension: ExtensionEntryData) {
  const settings = SettingsManager(socket, {
    extensionName: extension.name,
    configFile: extension.configPath + 'config.json',
    configVersion: CONFIG_VERSION,
    definitions: SettingDefinitions,
  });

  const context: Context = {
    api: API(socket),
    logger: socket.logger,
    getExtSetting: settings.getValue,
    extension,
  };

  const transporter = NodeMailer(context);
  const messageCollector = MessageCollector(transporter, context);

  const isSmtpSetting = (key: string) => key.indexOf('smtp_') === 0;

  extension.onStart = async (sessionInfo: any) => {
    const hostname = sessionInfo.system_info.hostname;
    settings.onValuesUpdated = (updatedValues: Record<string, unknown>) => {
      if (Object.keys(updatedValues).find(isSmtpSetting)) {
        transporter.reset(hostname);
      }
    };

    await settings.load();
    socket.logger.verbose('Settings were loaded');

    if (settings.getValue('hub_message_mode') != HubMessageCollectorMode.DISABLED) {
      socket.addListener('hubs', 'hub_message', messageCollector.onHubMessage);
    }

    if (settings.getValue('private_chat_message_mode') != PrivateChatCollectorMode.DISABLED) {
      socket.addListener('private_chat', 'private_chat_message', messageCollector.onPrivateMessage);
    }
  };

  extension.onStop = () => {
    messageCollector.stop();
  };
}
