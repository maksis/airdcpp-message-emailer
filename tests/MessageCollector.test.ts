import { SessionId } from '../src/types';

import { describe, test, expect, beforeEach, afterEach, vi } from 'vitest';
import { Hub1, Hub2, HubUser1, PrivateChat1, PrivateChat2, User1Message1, User2Message1 } from './mocks/mock-entities';

import { MessageCollector } from '../src/messages/MessageCollector';
import { Transporter } from '../src/transport/NodeMailer';
import { createMockContext } from './mocks/mock-context';
import { HubMessageCollectorMode, PrivateChatCollectorMode } from '../src/settings';


describe('MessageCollector', () => {
  const getTimezoneOffset = Date.prototype.getTimezoneOffset;

  beforeEach(() => {
    Date.prototype.getTimezoneOffset = () => {
      return -180;
    }
  });

  afterEach(() => {
    Date.prototype.getTimezoneOffset = getTimezoneOffset;
  });

  const createMockTransport = () => {
    const transport: Transporter = {
      sendMail: vi.fn(),
      reset: vi.fn(),
    };

    return transport;
  }

  const PrivateChatSessionInfoFetcher = (sessionId: SessionId) => {
    return Promise.resolve(sessionId === HubUser1.cid ? PrivateChat1 : PrivateChat2);
  };

  const HubSessionInfoFetcher = (sessionId: SessionId) => {
    return Promise.resolve(sessionId === Hub1.id.toString() ? Hub1 : Hub2);
  };

  test('should collect messages', async () => {
    const apiOverrides = {
      postEvent: vi.fn(),
      getPrivateChatSession: PrivateChatSessionInfoFetcher,
      getHubSession: HubSessionInfoFetcher,
    };

    const mockTransport = createMockTransport();
    const context = createMockContext({ apiOverrides });
    const messageCollector = MessageCollector(mockTransport, context);

    // Add messages
    await messageCollector.onHubMessage(User1Message1, Hub1.id);
    await messageCollector.onPrivateMessage(User2Message1, PrivateChat1.id);

    const hubMessages = messageCollector.getHubSessionMessages(Hub1.id);
    expect(hubMessages).toEqual([User1Message1]);

    const privateMessages = messageCollector.getPrivateChatSessionMessages(PrivateChat1.id);
    expect(privateMessages).toEqual([User2Message1]);

    expect(messageCollector.hasPendingFlush()).toBe(true);

    // Flush
    await messageCollector.flush();

    expect(messageCollector.hasPendingFlush()).toBe(false);
    expect(mockTransport.sendMail).toHaveBeenCalledTimes(1);
    expect(mockTransport.sendMail).toMatchSnapshot();
  });

  test('should ignore messages', async () => {
    const apiOverrides = {
      postEvent: vi.fn(),
      getHubSession: HubSessionInfoFetcher,
    };

    const settingOverrides = {
      hub_message_mode: HubMessageCollectorMode.DISABLED,
      private_chat_message_mode: PrivateChatCollectorMode.DISABLED,
    }

    const mockTransport = createMockTransport();
    const context = createMockContext({ apiOverrides, settingOverrides });
    const messageCollector = MessageCollector(mockTransport, context);

    // Add messages
    await messageCollector.onHubMessage(User1Message1, Hub1.id);
    await messageCollector.onPrivateMessage(User2Message1, PrivateChat1.id);

    // Nothing should exist
    expect(messageCollector.hasPendingFlush()).toBe(false);
  });
});
