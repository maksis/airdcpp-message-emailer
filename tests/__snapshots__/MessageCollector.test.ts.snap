// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MessageCollector > should collect messages 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "------- Private messages -------

-- [100]User1 --

[19:51:22] <[100]User2> User 2 message 1


------- Hub messages -------

-- Hub 1 --

[19:34:42] <[100]User1> User 1 message 1


",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
